release_notes/model.py

```diff
@@ -0,0 +1,228 @@
+
+    def __hash__(self):
+        return hash(self.get_identifier())
+
+    def __eq__(self, other):
+        if isinstance(other, ReleaseNote):
+            return self.__eq__(other.source_block)
+        if isinstance(other, SourceBlock):
+            return hash(other) == hash(self)
+        return False
+
+
+pattern = re.compile(r'\x60{3}(?P<category>\w+)\s+(?P<target_group>\w+)\n(?P<note>.+?)\n\x60{3}',
+                     flags=re.DOTALL | re.IGNORECASE | re.MULTILINE)
+
+
+def iter_source_blocks(content: str) -> typing.Generator[SourceBlock, None, None]:
+    ''' Searches for code blocks in release note notation and returns all found.
+    Only valid note blocks are returned, which means that the format has been followed.
+    However, it does not check if the category / group exists.
+
+    :param content: the content to look for release notes in
+    :return: a list of valid note blocks
+    '''
+    for res in pattern.finditer(content.replace('\r\n', '\n')):
+        try:
+            block = SourceBlock(category=res.group('category'),
+                                target_group=res.group('target_group'),
+                                note_message=res.group('note'))
+            if block.has_content():
+                yield block
+        except IndexError:
+            # group not found, ignore
+            continue
+
+
+@dataclasses.dataclass
+class ReleaseNote:
+    source_commit: Commit  # the commit where the release notes were initially gathered from
+    source_block: SourceBlock
+
+    raw_body: str  # the raw body of the commit / pull request
+    author: typing.Optional[Author]  # the author of the commit / pull request
+    reference: Reference
+
+    source_component: gci.componentmodel.Component = dataclasses.field(compare=False)
+    is_current_repo: bool
+    from_same_github_instance: bool
+
+    def __hash__(self):
+        return hash(self.source_block.get_identifier())
+
+    def __eq__(self, other) -> bool:
+        return self.source_block.__eq__(other)
+
+    def get_author_str(self) -> str:
+        return str(self.author)
+
+    def get_reference_str(self) -> str:
+        return f'{self.reference.type.identifier}{self.reference.get_identifier()}'
+
+    def to_block_str(self) -> str:
+        return '```{category} {group} {src_repo} {ref} {author}\n{message}\n```'.format(
+            category=self.source_block.category,
+            group=self.source_block.target_group,
+            src_repo=self.source_component.name,
+            ref=self.get_reference_str(),
+            author=self.author.username or self.author.display_name.replace(' ', '-'),
+            message=self.source_block.note_message
+        )
+
+
+def create_release_note_obj(
+        block: SourceBlock,
+
+        source_commit: Commit,
+        raw_body: str,
+        author: typing.Optional[Author],
+
+        targets: typing.Union[git.Commit, github3.pulls.ShortPullRequest],
```

Comment:
plural form typically implies that a sequence/iterable is expected. rather name this `target`?